### PR TITLE
Safety checks, Retain component type in ExitStruct, miscellaneous build updates

### PR DESCRIPTION
**Pull request by rtgill82**
- Added initialization guards to prevent functions acting on components from being called when `libnewt` is uninitialized.
- Retain component Widget types so that specific Widgets may be returned with `ExitStruct`.
- Add bunder Gemfile
- Allow CFLAGS to be specified in environment.

### CODE REVIEW

**Comment by rtgill82** [comment]
I think I may be able to add a few more tests. So give me a bit more time before accepting this pull request.

**Comment by lzap** [comment]
Hey, I am glad to see this. This binding is not in the best shape, it has some memory issues here and there. We maintain this because of couple of screens we have, so we are not using ton of features. Maintainers welcomed!

Let me know when this is ready, I will go ahead and run tests and test our app on this change. There is no CI.

Looks like you accidentally created a PR from your master: `rtgill82:master` FYI.

**Comment by rtgill82** [comment]
_Looks like you accidentally created a PR from your master_

My mistake. No one else has forked this though, and there haven't been any changes since the last merge, so it shouldn't be too much of an issue. I'll keep it in mind for the future though.

**Comment by rtgill82** [comment]
Ok, I think I'm done with this pull request and I completed everything I hoped to.

**Comment by rtgill82** [comment]
Fixed the issue with callbacks being freed by the GC. I think this is ready for merge now.

**Comment by lzap** [comment]
Thanks, so before I checkout your branch I want to test out 1.0 more, it looks like I haven't got to releasing the final 1.0 version only pre-release. But I am hitting:

```
[lzap@nuc ruby-newt]$ rake
install -c tmp/x86_64-linux/ruby_newt/2.7.2/ruby_newt.so lib/ruby_newt/ruby_newt.so
cp tmp/x86_64-linux/ruby_newt/2.7.2/ruby_newt.so tmp/x86_64-linux/stage/lib/ruby_newt/ruby_newt.so
[lzap@nuc ruby-newt]$ ruby -Ilib -Itmp/x86_64-linux/stage/lib/ruby_newt examples/test_method/RadioButton.rb
Traceback (most recent call last):
	5: from examples/test_method/RadioButton.rb:4:in `<main>'
	4: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
	3: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
	2: from /home/lzap/work/ruby-newt/lib/newt.rb:1:in `<top (required)>'
	1: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require': /home/lzap/work/ruby-newt/lib/ruby_newt/ruby_newt.so: undefined symbol: newtDefaultColorPalette - /home/lzap/work/ruby-newt/lib/ruby_newt/ruby_newt.so (LoadError)
```

Do you know what can be wrong? I am on Fedora 35 with newt-0.52.21-11.fc35.x86_64

```
$ grep newtDefaultColorPalette /usr/include/newt.h
extern const struct newtColors newtDefaultColorPalette;
```

File: ext/ruby_newt/ruby_newt.c [hunk 1]
```diff
@@ -184,28 +184,21 @@ static VALUE rb_ext_ColorSetCustom(VALUE self, VALUE id)
...
-
-  newtInit();
-  memcpy(&newtColors, &newtDefaultColorPalette, sizeof(struct newtColors));
-  rb_ext_Widget_Hash = rb_hash_new();
-  rb_gc_register_address(&rb_ext_Widget_Hash);
-  initialized = Qtrue;
+  rb_ext_Screen_Init();
+  newtCls();
   return Qnil;
 }
```
Comment (lzap) [inline ext/ruby_newt/ruby_newt.c:203]: When called twice, cls is called twice now, probably expected behavior.

Unsure about rb_gc_register_address change tho, I do not see a change for the unregister counterpart in this commit (probably somewhere else)?

**Comment by lzap** [comment]
Are there any new 0.9 to 1.0 incompatible changes? Since 1.0 hasn't been released yet, I think we can do pre2 and then push out 1.0.

**Comment by rtgill82** [comment]
The majority of methods can no longer be called outside of `Newt::Screen.init`, `Newt::Screen.finish` pairs. I had to update the example scripts because a lot of them printed method results after `Newt::Screen.finish` was called. It's best to assign results to a variable to be manipulated later.

Equality functions do not raise the "uninitialized" exception, so it's possible to inadvertently compare components that have been instantiated between different calls of `Newt::Screen.init`. It's also possible to call methods on components that have already been destroyed if `Newt::Screen.init` is called again after `Newt::Screen.finish`. I'm not sure if this should really be worried about? Perhaps I'll figure out how to fix it later.

**Comment by rtgill82** [comment]
Hmm... I'm not sure. It seems that RPM revision doesn't have that symbol.

`zsh 1066 % pwd`
`/home/robert/tmp/build/rpm/newt-0.52.21-11.fc35.x86_64/usr/lib64`
`zsh 1067 % nm libnewt.so.0.52.21 | grep -i Default`
`nm: libnewt.so.0.52.21: no symbols`

Where as on my machine running Gentoo with newt package
`dev-libs/newt-0.52.21-r1` I get:

`zsh 1072 % nm /usr/lib64/libnewt.so.0.52.21 | grep -i Default`
`00000000000156e0 b default_keyreader_buf`
`0000000000014c40 D newtDefaultColorPalette`
`0000000000006b38 T newtDefaultEventHandler`
`0000000000006de0 T newtDefaultMappedHandler`
`0000000000006dd9 T newtDefaultPlaceHandler`

**Comment by rtgill82** [comment]
It looks like the `-flto=auto` and `-ffat-lto-objects` `CFLAGS` that Fedora includes by default strip out `newtDefaultColorPalette`.

**Comment by rtgill82** [comment]
Never mind, the symbol is there if I run `nm` with the `-D` option. I don't know why the linker isn't picking it up in your case.

`zsh/2 1070 % nm -D /usr/lib64/libnewt.so.0.52.21 | grep Default`
`0000000000016c40 D newtDefaultColorPalette@@NEWT_0.52`
`0000000000005910 T newtDefaultEventHandler@@NEWT_0.52`
`0000000000005930 T newtDefaultMappedHandler@@NEWT_0.52`
`0000000000005920 T newtDefaultPlaceHandler@@NEWT_0.52`

**Comment by rtgill82** [comment]
Meant to paste this, which is the Fedora 35 binary.

`zsh/2 1071 % nm -D ./libnewt.so.0.52.21 | grep Default`
`0000000000016820 D newtDefaultColorPalette@@NEWT_0.52`
`0000000000006110 T newtDefaultEventHandler@@NEWT_0.52`
`0000000000006130 T newtDefaultMappedHandler@@NEWT_0.52`
`0000000000006120 T newtDefaultPlaceHandler@@NEWT_0.52`

**Comment by rtgill82** [comment]
I installed Fedora 35 on a VM and wasn't able to reproduce the issue.
I'm running packages `ruby-3.0.2-151.fc35.x86_64` and `newt-0.52.21-11.fc35.x86_64`.

**Comment by rtgill82** [comment]
I note that you're using ruby-2.7. I don't know where `rake-compiler` grabs its CFLAGS from, but I assume it uses the same that ruby was built with. Maybe the ruby-2.7 CFLAGS are incompatible with the current newt package. This is how it compiles on my Fedora 35 VM.

`gcc -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I../../../../ext/ruby_newt   -fPIC -O2  -fexceptions -g -grecord-gcc-switches
 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-pro
tector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
 -fPIC -m64 -o ruby_newt.o -c ../../../../ext/ruby_newt/ruby_newt.c`
`gcc -shared -o ruby_newt.so ruby_newt.o -L. -L/usr/lib64 -L. -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-
ld -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld
  -m64  -lruby -lnewt -lslang  -lm   -lc`

**Comment by rtgill82** [comment]
Added LDFLAGS to the build script and made a force-push.

**Comment by lzap** [comment]
Sorry for the delay, correct, I have Ruby 2.7.2 from rbenv and it is very likely that I compiled it on older Fedora before upgrade. That's something I should definitely avoid. I work on a monster Ruby on Rails application and catching up with Ruby and RoR versions is a challenge and we are typically behind.

I am reinstalling 2.7.5 from scratch now and I will test your patch once again. If it works fine, I think let's just push 1.0 out of doors and fix bugs, there is no need to wait.

**Comment by lzap** [comment]
And indeed it compiles fine:

```
[lzap@nuc ruby-newt]$ gem install rake-compiler
Fetching rake-compiler-1.1.9.gem
Successfully installed rake-compiler-1.1.9
1 gem installed
[lzap@nuc ruby-newt]$ rake compile
mkdir -p tmp/x86_64-linux/ruby_newt/2.7.5
cd tmp/x86_64-linux/ruby_newt/2.7.5
/home/lzap/.rbenv/versions/2.7.5/bin/ruby -I. -r.rake-compiler-siteconf.rb ../../../../ext/ruby_newt/extconf.rb
checking for SLsmg_refresh() in -lslang... yes
checking for newtInit() in -lnewt... yes
creating Makefile
cd -
cd tmp/x86_64-linux/ruby_newt/2.7.5
/usr/bin/gmake
compiling ../../../../ext/ruby_newt/ruby_newt.c
linking shared-object ruby_newt/ruby_newt.so
cd -
/usr/bin/gmake install target_prefix=
/usr/bin/install -c -m 0755 ruby_newt.so /home/lzap/work/ruby-newt/lib/ruby_newt
cp tmp/x86_64-linux/ruby_newt/2.7.5/ruby_newt.so tmp/x86_64-linux/stage/lib/ruby_newt/ruby_newt.so
```

**Review by lzap — APPROVED** [review]
Apologies for the delay, great work. Works fine, I ported our application.

I will push out a new major version.

**Comment by lzap** [comment]
Thanks again! I've updated the README with some information from this change, if you find more things please add them.

```
[lzap@nuc ruby-newt]$ gem build newt.gemspec
  Successfully built RubyGem
  Name: newt
  Version: 1.0.0
  File: newt-1.0.0.gem

[lzap@nuc ruby-newt]$ gem push newt-1.0.0.gem
Pushing gem to https://rubygems.org...
You have enabled multi-factor authentication. Please enter OTP code.
Code:   516055
Successfully registered gem: newt (1.0.0)
```

**Pull request merged**